pulp_container/tests/functional/api/test_push_content.py

```diff
@@ -6,49 +6,145 @@
+            except exceptions.CalledProcessError:
+                pass
+
+        cls.registry.pull(image_path)
+
+        if user:
+            cls.registry.logout(cls.registry_name)
 
-    def test_push_using_registry_client(self):
-        """Test push with official registry client"""
-        registry.raise_if_unsupported(unittest.SkipTest, "Test requires podman/docker")
-        registry_name = urlparse(cfg.get_base_url()).netloc
-        local_url = "/".join([registry_name, "foo/bar:1.0"])
-        # Be sure to not being logged in
-        try:
-            registry.logout(registry_name)
-        except exceptions.CalledProcessError:
-            pass
-        # Pull an image with large blobs
-        registry.pull("centos:7")
+    @classmethod
+    def tag_and_push(cls, image_path, local_url, user):
+        """
+        Tag and push an image to Pulp registry using specified user.
+        """
         # Tag it to registry under test
-        registry.tag("centos:7", local_url)
-        # Try to push without permission
-        with self.assertRaises(exceptions.CalledProcessError):
-            registry.push(local_url)
+        cls.registry.tag(image_path, local_url)
         # Log in
-        registry.login("-u", "admin", "-p", "password", registry_name)
+        cls.registry.login("-u", user["username"], "-p", user["password"], cls.registry_name)
         # Push successfully
-        registry.push(local_url)
-        # Pull while logged in
-        registry.pull(local_url)
+        cls.registry.push(local_url)
         # Untag local copy
-        registry.rmi(local_url)
-        # Pull while logged out
-        registry.pull(local_url)
-        # cleanup
-        distribution = distributions_api.list(name="foo/bar").results[0]
-        self.addCleanup(distributions_api.delete, distribution.pulp_href)
+        cls.registry.rmi(local_url)
+
+        cls.registry.logout(cls.registry_name)
+
+    def test_push_using_registry_client_admin(self):
+        """Test push with official registry client and logged in as admin."""
+        image_path = "centos:7"
+        local_url = "/".join([self.registry_name, "foo/bar:1.0"])
+
+        self.tag_and_push(image_path, local_url, self.user_admin)
+        self._pull(local_url, self.user_admin)
+
+        # cleanup, namespace removal also removes related distributions
+        namespace = self.namespace_api.list(name="foo").results[0]
+        self.addCleanup(self.namespace_api.delete, namespace.pulp_href)
+
+    def test_push_without_login(self):
+        """Test that one can't push without being logged in."""
+        local_url = "/".join([self.registry_name, "foo/bar:1.0"])
+
+        # Try to push without permission
+        with self.assertRaises(exceptions.CalledProcessError):
+            self.registry.push(local_url)
+
+    def test_push_with_dist_perms(self):
+        """
+        Test that it's enough to have container distribution perms to perform push.
```

Comment:
this test shows that it is enough to have container distribution + namespace perms to push.
It covers the case when the namespace did not exist before push.

Can we add one more test what shows it is enough to have container dists perms + manage_dists_within_the_namespace when there is already an existing namespace? That would require to create a user_collaborator

Reply:
on the other hand, this push workflow has a lot of combinations, we can also add more extensive tests in the next PR pushes.